File: src/main/java/com/improve_future/s3explorer/service/AwsS3Service.java
package com.improve_future.s3explorer.service;

import com.amazonaws.services.s3.model.Bucket;
import com.amazonaws.services.s3.model.ListObjectsRequest;
import com.amazonaws.services.s3.model.ObjectListing;
import com.improve_future.s3explorer.config.AwsS3Config;
import com.improve_future.s3explorer.gateway.AwsS3Gateway;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by kenji on 16/09/17.
 */
public class AwsS3Service {
//    int equals(AwsS3Service awsS3Service) {
//
//    }

//    private static List<AwsS3Config> configList = new ArrayList<AwsS3Config>();
//    private static List<AwsS3Service> instances = new ArrayList<AwsS3Service>();

    private static AwsS3Service instance;
    private AwsS3Gateway awsS3Gateway;

    public static AwsS3Service getInstance(AwsS3Config awsS3Config) {
        if (instance == null) {
            instance = new AwsS3Service(awsS3Config);
        }
        return instance;
    }

    private AwsS3Service(AwsS3Config config) {
        awsS3Gateway = new AwsS3Gateway(config);
    }

    public List<Bucket> findAllBuckets() {
        return awsS3Gateway.findAllBuckets();
    }

    public ObjectListing findAllObjects(
            String bucketName,
            String prefix) {
        return this.findAllObjects(bucketName, prefix, "/");
    }

    public ObjectListing findAllObjects(
            String bucketName,
            String prefix,
            String delimiter) {
        ListObjectsRequest request = new ListObjectsRequest();
        request.setBucketName(bucketName);
        request.setPrefix(prefix);
        request.setDelimiter(delimiter);
        return awsS3Gateway.findAllFolders(request);
    }
}

File: src/main/java/com/improve_future/s3explorer/service/AwsS3Service.kt
package com.improve_future.s3explorer.service

import com.amazonaws.services.s3.model.*
import com.improve_future.s3explorer.config.AwsS3Config
import com.improve_future.s3explorer.gateway.AwsS3Gateway
import com.improve_future.s3explorer.gateway.FileSystemGateway

class AwsS3Service private constructor(config: AwsS3Config) {
    private val awsS3Gateway: AwsS3Gateway
    private val fileSystemGateway: FileSystemGateway

    init {
        awsS3Gateway = AwsS3Gateway(config)
        fileSystemGateway = FileSystemGateway()
    }

    fun findAllBuckets(): List<Bucket> {
        return awsS3Gateway.findAllBuckets()
    }

    fun findAllObjects(
            bucketName: String,
            prefix: String,
            delimiter: String = "/"): ObjectListing {
        return awsS3Gateway.findAllObjects(
                bucketName, prefix, delimiter)
    }

    fun deleteObject(
            bucketName: String,
            key: String) {
        return awsS3Gateway.deleteObject(
                bucketName, key)
    }

    fun downloadObject(
            bucketName: String,
            key: String) {
        val stream = awsS3Gateway.findObjectStream(bucketName, key)
        fileSystemGateway.save(stream, bucketName + "/" + key)

    }

    companion object {
        //    int equals(AwsS3Service awsS3Service) {
        //
        //    }

        //    private static List<AwsS3Config> configList = new ArrayList<AwsS3Config>();
        //    private static List<AwsS3Service> instances = new ArrayList<AwsS3Service>();

        private var instance: AwsS3Service? = null

        fun getInstance(awsS3Config: AwsS3Config): AwsS3Service {
            if (instance == null) {
                instance = AwsS3Service(awsS3Config)
            }
            return instance!!
        }
    }
}
